zerver/lib/attachments.py

```diff
@@ -160,6 +162,24 @@ def validate_attachment_request(
         # can access it.
         return True, attachment
 
+    message_channel_ids = set()
+    for message in messages:
+        if message.is_stream_message():
+            message_channel_ids.add(message.recipient.type_id)
+
+    # We only fetch channels with public history, since those are the
+    # only channels that user groups can grant content access to.
+    message_channels_with_public_history = Stream.objects.filter(
+        id__in=message_channel_ids, history_public_to_subscribers=True
+    )
```

Comment:
I think if you do `message.recipient_id` and `recipient_id__in=message_recipient_ids`, you can avoid needing to `.select_related("recipient")`.

Reply:
I tried this, it does not make a difference on the query count, since we will ultimately need to fetch the recipients when fetching subscriptions. A separate recipient query is called only once in the query logs as seen here: https://pastebin.com/Ltq30jgS